internal/check/bottom_up.go

```diff
@@ -0,0 +1,590 @@
+package check
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"slices"
+
+	"github.com/sourcegraph/conc/panics"
+
+	openfgav1 "github.com/openfga/api/proto/openfga/v1"
+	authzGraph "github.com/openfga/language/pkg/go/graph"
+
+	"github.com/openfga/openfga/internal/concurrency"
+	"github.com/openfga/openfga/internal/condition"
+	"github.com/openfga/openfga/internal/iterator"
+	"github.com/openfga/openfga/internal/modelgraph"
+	"github.com/openfga/openfga/pkg/storage"
+	"github.com/openfga/openfga/pkg/tuple"
+)
+
+type strategyKind int64
+
+const (
+	normal strategyKind = iota
+	recursive
+)
+
+type bottomUpHandler func(context.Context, []storage.Iterator[string], chan<- *iterator.Msg)
+
+type bottomUp struct {
+	model     *modelgraph.AuthorizationModelGraph
+	datastore storage.RelationshipTupleReader
+	strategy  strategyKind
+}
+
+func newBottomUp(model *modelgraph.AuthorizationModelGraph, ds storage.RelationshipTupleReader) *bottomUp {
+	return &bottomUp{
+		model:     model,
+		datastore: ds,
+		strategy:  normal,
+	}
+}
+
+func newBottomUpRecursive(model *modelgraph.AuthorizationModelGraph, ds storage.RelationshipTupleReader) *bottomUp {
+	return &bottomUp{
+		model:     model,
+		datastore: ds,
+		strategy:  recursive,
+	}
+}
+
+func (s *bottomUp) setWildcardIntersectionSetup(ctx context.Context, req *Request, edges []*authzGraph.WeightedAuthorizationModelEdge) (chan *iterator.Msg, error) {
+	shortcircuitIntersection := false
+
+	// if there is one edge that does not contain the wildcard, we can short-circuit
+	for _, edge := range edges {
+		if !slices.Contains(edge.GetWildcards(), req.GetUserType()) {
+			shortcircuitIntersection = true
+			break
+		}
+	}
+
+	if !shortcircuitIntersection {
+		return s.setOperationSetup(ctx, req, resolveIntersection, edges)
+	}
+
+	iterChan := make(chan *iterator.Msg)
+	close(iterChan)
+	return iterChan, nil
+}
+
+// setOperationSetup returns a channel with a number of elements that is >= the number of children.
+// Each element is an iterator.
+// The caller must wait until the channel is closed.
+func (s *bottomUp) setOperationSetup(ctx context.Context, req *Request, resolver bottomUpHandler, edges []*authzGraph.WeightedAuthorizationModelEdge) (chan *iterator.Msg, error) {
+	iterChans := make([]storage.Iterator[string], 0, len(edges))
+	for _, edge := range edges {
+		if _, ok := edge.GetWeight(req.GetUserType()); !ok {
+			continue
+		}
+		producerChan, err := s.resolveEdge(ctx, req, edge)
+		if err != nil {
+			return nil, err
+		}
+		iterChans = append(iterChans, iterator.FromChannel(producerChan))
+	}
+
+	out := make(chan *iterator.Msg, len(iterChans))
+	go func() {
+		recoveredError := panics.Try(func() {
+			resolver(ctx, iterChans, out)
+		})
+
+		if recoveredError != nil {
+			fmt.Println(recoveredError.AsError())
```

Comment:
Instead of fmt.Println - should this be via logger?  (maybe passed in logger to bottomUp)